nuxeo-core-binarymanager-s3/src/main/java/org/nuxeo/ecm/core/storage/sql/S3BinaryManager.java

```diff
@@ -428,6 +448,58 @@ protected FileStorage getFileStorage() {
         return new S3FileStorage();
     }
 
+    @Override
+    public String writeBlob(Blob blob) throws IOException {
+        // Attempt to do S3 Copy if the Source Blob provider is also S3
+        if (ManagedBlob.class.isAssignableFrom(blob.getClass())) {
+            ManagedBlob managedBlob = (ManagedBlob) blob;
+            BlobProvider blobProvider = Framework.getService(BlobManager.class).getBlobProvider(managedBlob.getProviderId());
+
+            if (this != blobProvider) {
+                if (this.getClass().isAssignableFrom(blobProvider.getClass())) {
+                    S3BinaryManager sourceBlobProvider = (S3BinaryManager) blobProvider;
+
+                    String digest = managedBlob.getKey();
+                    int colon = digest.indexOf(':');
+                    if (colon >= 0) {
+                        digest = digest.substring(colon + 1);
+                    }
+
+                    try {
+                        ObjectMetadata targetMetadata = amazonS3.getObjectMetadata(bucketName, digest);
+                        // The target blob already exists nothing to do.
+                        return targetMetadata.getETag();
+                    } catch (AmazonS3Exception e) {
+                        // Object does not exist, just continue
+                    }
+
+                    ObjectMetadata sourceMetadata;
+
+                    try {
+                        sourceMetadata = amazonS3.getObjectMetadata(sourceBlobProvider.bucketName, digest);
+                    } catch (AmazonS3Exception e) {
+                        throw new NuxeoException(MessageFormat.format("Source blob does not exists: s3://{0}/{1}", sourceBlobProvider.bucketName,
+                                digest, blob), e);
+                    }
+
+                    try {
+                        ObjectMetadata result;
+                        if (sourceMetadata.getContentLength() >= FIVE_GB) {
+                            result = AWSUtils.copyBigFile(amazonS3, sourceMetadata, sourceBlobProvider.bucketName, digest, bucketName, digest, true);
+                        } else {
+                            result = AWSUtils.copyFile(amazonS3, sourceMetadata, sourceBlobProvider.bucketName, digest, bucketName, digest, true);
+                        }
+
+                        return result.getETag();
+                    } catch (Exception e) {
```

Comment:
Could you reduce the catch ? (ie: `Exception` is a bit too wide

Reply:
It's a bit hard, since AWS SDK can throw lots of Exception and `copyBigFile` already throws NuxeoException